packages/flutter/lib/src/cupertino/date_picker.dart

```diff
@@ -448,7 +448,7 @@ class _CupertinoDatePickerDateTimeState extends State<CupertinoDatePicker> {
       date.year,
       date.month,
       date.day,
-      selectedHour + selectedAmPm * 12,
+      selectedHour % 12 + selectedAmPm * 12,
```

Comment:
If use24hFormat is true then selectedAmPm is 0 and the %12 isn't correct.

Reply:
`selectedAmPm * 12 + (widget.use24hFormat ? selectedHour : selectedHour % 12)`
Too long?

Reply:
Actually this is a little clearer about intent:
`widget.use24hFormat ? selectedHour : selectedHour % 12 + selectedAmPm * 12`

Reply:
That looks good. 

If there isn't another test that checks this, we'll need a new test that checks this logic in the 24hFormat case.  Not to prove that your change is correct (looks fine to me) but to defend against some future change introducing a regression.